Service operators can easily identify claims that are close to their deadline

robbiepaul opened the pull request:
We tell users they will hear back on the outcome of their claim "within 6 weeks". To help Service Operators ensure they check claims within the 6 weeks, they should be able to easily see when a claim is nearing the end of this period, and when it has passed it.

## Screenshots

<img width="1785" alt="Screenshot 2019-10-14 at 14 44 29" src="https://user-images.githubusercontent.com/2804149/66756103-364a0500-ee91-11e9-96de-52b476fa9ecc.png">
<img width="1785" alt="Screenshot 2019-10-14 at 14 44 38" src="https://user-images.githubusercontent.com/2804149/66756104-364a0500-ee91-11e9-9f9c-80a8f3dc9ba5.png">
<img width="1785" alt="Screenshot 2019-10-14 at 14 44 48" src="https://user-images.githubusercontent.com/2804149/66756105-364a0500-ee91-11e9-8d47-b9c88ce28fd1.png">

robbiepaul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Service operators can easily identify claims that are close to the SLA for claim checking](https://trello.com/c/OTJs4W1P/813-service-operators-can-easily-identify-claims-that-are-close-to-the-sla-for-claim-checking)

erbridge commented:
Does the deadline column in the list view need a heading to give it context?

erbridge commented on app/models/claim.rb:
```diff
@@ -39,6 +39,7 @@ class Claim < ApplicationRecord
     building_society_roll_number: true,
     payroll_run_id: false,
   }.freeze
+  DEADLINE = 6.weeks
```
Think we can derive the deadline text we display to claimants from this, too?

erbridge replied:
Is there not a `to_weeks` method or something?

erbridge commented on spec/helpers/admin/claims_helper_spec.rb:
```diff
@@ -94,8 +94,21 @@
       ])
     end
+
+    context "when the claim is approaching it's deadline" do
```
```suggestion
    context "when the claim is approaching its deadline" do
```

robbiepaul commented:
Yeah I was on the fence about that, I guess because it doesn't always show I didn't want to have it as `Weekdays until deadline` but perhaps something like `Deadline warning` would work

erbridge commented:
Or only show the column at all if there are any? Or modify the tect on the button to explain what it refers to?

tekin commented on app/helpers/admin/claims_helper.rb:
```diff
@@ -50,5 +51,21 @@ def display_school(school)
+
+    private
+
+    def weekdays_between(first_date, second_date)
```
What if this method was updated to return a negative value when the second date was before the first date? Would that simplify the `deadline_warning` method?

tekin commented on app/models/claim.rb:
```diff
@@ -39,6 +39,7 @@ class Claim < ApplicationRecord
     building_society_roll_number: true,
     payroll_run_id: false,
   }.freeze
+  DEADLINE = 6.weeks
```
`DEADLINE` is a very vague term. Can we give it a more specific name? `CHECKING_SLA_WINDOW` or something?

tekin commented on app/views/admin/claims/index.html.erb:
```diff
@@ -15,13 +15,17 @@
             <th scope="col" class="govuk-table__header">Reference</th>
             <th scope="col" class="govuk-table__header">Applicant Name</th>
             <th scope="col" class="govuk-table__header"></th>
+            <th scope="col" class="govuk-table__header">Deadline</th>
```
```suggestion
            <th scope="col" class="govuk-table__header">Checking deadline</th>
```

tekin commented on app/models/claim.rb:
```diff
@@ -140,6 +141,22 @@ def payroll_gender_missing?
+    submitted? && check.nil?
+  end
+
+  def approaching_deadline?
```
I can't help but wonder, what if we had a `days_until_checking_deadline` that returns either a positive or negative number and used that to determine where the claim is on the schedule and also to generate warning markup? Would that make things simpler?

robbiepaul replied:
Yeah it would make it a lot simpler. It would mean that I can remove three methods from the claim model and simplify the `deadline_warning` helper at the same time

tekin commented on app/models/claim.rb:
```diff
@@ -140,6 +141,22 @@ def payroll_gender_missing?
+    awaiting_check? && Time.zone.now > deadline_date
+  end
+
+  def deadline_date
```
As with the constant above, would be good to give these methods more specific names, e.g. `checking_deadline_date`

tekin commented on app/views/admin/page/index.html.erb:
```diff
@@ -1,17 +1,24 @@
-    <p class="govuk-body govuk-!-font-size-36"><%= Claim.awaiting_checking.count %></p>
+  <div class="govuk-grid-column-one-half">
+    <h2 class="govuk-heading-m">Claims approaching deadline</h2>
+    <p class="govuk-body govuk-!-font-size-36"><%= @claims_awaiting_checking.select(&:approaching_deadline?).count %></p>
```
This is going to load all the claims into memory to perform the count. Is there a way to calculate this number, and the number below using a query instead and set them as instance variables in the controller action?

robbiepaul commented on app/controllers/admin/page_controller.rb:
```diff
@@ -1,6 +1,7 @@
 module Admin
   class PageController < BaseAdminController
     def index
+      @claims_awaiting_checking = Claim.includes(:check).awaiting_checking
```
Yeah I thought that too, but I don't think it does the eager loading.

Before `.includes(:check)`:

```
11:38:18 web.1    |   Claim Load (26.3ms)  SELECT "claims".* FROM "claims" LEFT OUTER JOIN "checks" ON "checks"."claim_id" = "claims"."id" WHERE "claims"."submitted_at" IS NOT NULL AND "checks"."claim_id" IS NULL
11:38:18 web.1    |   ↳ app/views/admin/page/index.html.erb:15
11:38:18 web.1    |   Check Load (24.7ms)  SELECT  "checks".* FROM "checks" WHERE "checks"."claim_id" = $1 LIMIT $2  [["claim_id", "5a2318f7-b77e-48ee-9b92-bf5b95725f9f"], ["LIMIT", 1]]
11:38:18 web.1    |   ↳ app/models/claim.rb:145
11:38:18 web.1    |   Check Load (28.0ms)  SELECT  "checks".* FROM "checks" WHERE "checks"."claim_id" = $1 LIMIT $2  [["claim_id", "80a4a367-16db-4e41-939b-d46da911980d"], ["LIMIT", 1]]
11:38:18 web.1    |   ↳ app/models/claim.rb:145
11:38:18 web.1    |   Check Load (26.5ms)  SELECT  "checks".* FROM "checks" WHERE "checks"."claim_id" = $1 LIMIT $2  [["claim_id", "31041254-edfa-4b07-b024-a87f3f888bee"], ["LIMIT", 1]]
11:38:18 web.1    |   ↳ app/models/claim.rb:145
```

After `.includes(:check)`:
```
11:42:50 web.1    |   ↳ app/views/admin/page/index.html.erb:9
11:42:50 web.1    |   Claim Load (26.4ms)  SELECT "claims".* FROM "claims" LEFT OUTER JOIN "checks" ON "checks"."claim_id" = "claims"."id" WHERE "claims"."submitted_at" IS NOT NULL AND "checks"."claim_id" IS NULL
11:42:50 web.1    |   ↳ app/views/admin/page/index.html.erb:15
11:42:50 web.1    |   Check Load (27.5ms)  SELECT "checks".* FROM "checks" WHERE "checks"."claim_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11) .....
```

However saying that, I'm going to be changing this action so it doesn't load all the claims in memory anyway

robbiepaul commented on app/helpers/admin/claims_helper.rb:
```diff
@@ -50,5 +51,21 @@ def display_school(school)
       sanitize(html, tags: %w[span a], attributes: %w[href class])
     end
+
+    def deadline_warning(claim)
```
I think this method is a lot simpler now that I've updated the code to have a `weekdays_until_checking_deadline`, it still has a guard clause to filter out claims still have more than 10 days left to be checked though.

robbiepaul commented:
@tekin Ready for another look

erbridge commented on app/assets/stylesheets/components/claim_deadline.scss:
```diff
@@ -0,0 +1,25 @@
+  letter-spacing: 0;
+  min-width: 4em;
+  text-align: center;
+}
```
Is this a generic style for tags (or whatever we're calling those things)? Does it make sense to make it a generic component that we import?

erbridge commented on app/models/claim.rb:
```diff
@@ -115,6 +116,8 @@ class Claim < ApplicationRecord
   scope :submitted, -> { where.not(submitted_at: nil) }
   scope :awaiting_checking, -> { submitted.left_outer_joins(:check).where(checks: {claim_id: nil}) }
   scope :approved, -> { joins(:check).where("checks.result" => :approved) }
+  scope :approaching_checking_deadline, -> { awaiting_checking.where("submitted_at < ?", CHECKING_DEADLINE.ago + 2.weeks).where.not("submitted_at < ?", CHECKING_DEADLINE.ago) }
```
Could we pull the `+ 2.weeks` out into another constant? What we consider to be "approaching" is a bit hidden.

erbridge commented on app/assets/stylesheets/components/claim_deadline.scss:
```diff
@@ -0,0 +1,25 @@
+@import "govuk-frontend-rails";
+
+%claim-deadline {
+  text-transform: lowercase;
```
Probably better to make sure we're actually using lowercase text so that copying it (and I guess screen readers?) don't have the raw text with uppercase letters.

robbiepaul replied:
Yeah the text is actually lowercase but the `govuk-tag` class makes it appear uppercase. I've change our override to be `text-transform: none`

erbridge replied:
Aha! Makes sense.

erbridge commented on app/helpers/admin/claims_helper.rb:
```diff
@@ -50,5 +51,22 @@ def display_school(school)
+
+    def checking_deadline_warning(claim)
+      weekdays_until_checking_deadline = weekdays_between(Date.today, claim.checking_deadline_date)
+      return if weekdays_until_checking_deadline > 10
```
Can we pull this `10` out into a constant so it's easier to follow the purpose of it (and see what it is)?

tekin replied:
Yeah, this value and the `2.weeks` in the scope are effectively the same thing, right?

erbridge commented on app/helpers/admin/claims_helper.rb:
```diff
@@ -50,5 +51,22 @@ def display_school(school)
+      weekdays = (dates_in_order.first..dates_in_order.last).select { |date| [0, 6].exclude?(date.wday) }.count
+      weekdays *= -1 if first_date > second_date
+      weekdays
+    end
```
I'm surprised Ruby doesn't have a `(date_a - date_b).to_days`.

erbridge commented on spec/helpers/admin/claims_helper_spec.rb:
```diff
@@ -114,4 +127,30 @@
+
+  describe "#checking_deadline_warning" do
+    subject { helper.checking_deadline_warning(claim) }
+    before { travel_to Time.zone.local(2019, 10, 11, 7, 0, 0) }
```
Is it important that it's a given date and not just frozen?

robbiepaul replied:
Hmm yea, I guess the spec could be a little clearer. It needs to be frozen because it doesn't count weekends. I'll add another context

tekin commented on app/helpers/admin/claims_helper.rb:
```diff
@@ -50,5 +51,22 @@ def display_school(school)
+
+    def weekdays_between(first_date, second_date)
+      dates_in_order = [first_date, second_date].sort
+      weekdays = (dates_in_order.first..dates_in_order.last).select { |date| [0, 6].exclude?(date.wday) }.count
```
Reckon we can simplify this line with [`Date.#on_weekday?`](https://api.rubyonrails.org/v5.1/classes/DateAndTime/Calculations.html#method-i-on_weekday-3F):

```suggestion
      weekdays = (dates_in_order.first..dates_in_order.last).count { |date| date.on_weekday? }
```

robbiepaul commented:
- Renamed 'checking deadline' to 'check deadline'
- Renamed `CHECK_DEADLINE_WARNING_POINT`
- Warning shows number of calendar days now, not weekdays

tekin commented on spec/features/admin_stats_spec.rb:
```diff
@@ -0,0 +1,29 @@
+    visit admin_root_path
+  end
+
+  scenario "Service operator is shown how many claims have been submitted" do
```
Can we combine these four scenarios into a single "Service operator is shown claim stats"? Each scenario has exactly the same expensive setup and is checking the same page. We only need to create the records, log the user in and render the page once to verify the expected behaviour. That'll make reduce the hit on our test suite run time a chunk.

tekin commented on app/models/claim.rb:
```diff
@@ -115,6 +117,8 @@ class Claim < ApplicationRecord
   scope :submitted, -> { where.not(submitted_at: nil) }
   scope :awaiting_checking, -> { submitted.left_outer_joins(:check).where(checks: {claim_id: nil}) }
   scope :approved, -> { joins(:check).where("checks.result" => :approved) }
+  scope :approaching_check_deadline, -> { awaiting_checking.where("submitted_at < ?", CHECK_DEADLINE.ago + CHECK_DEADLINE_WARNING_POINT).where.not("submitted_at < ?", CHECK_DEADLINE.ago) }
```
This could probably be expressed more simply with a "submitted_at < ? AND submitted_at > ?" but not a blocker.
